LSM: implement move table optimization for compaction

kprotty opened the pull request:
When the table can be moved directly between levels without reading values from the iterators, do so and skip the rest of the compaction process.

This is currently a prototype. The benchmark ends up hitting this case quite frequently and not really doing any ~~Grid~~ disk reads.

## Pre-merge checklist

Performance:

* [x] Compare `zig benchmark` on linux before and after this pr.
    ``` sh
    # benchmark results before
    1223 batches in 371.97 s
    load offered = 1000000 tx/s
    load accepted = 26883 tx/s
    batch latency p00 = 5 ms
    batch latency p10 = 37 ms
    batch latency p20 = 42 ms
    batch latency p30 = 46 ms
    batch latency p40 = 47 ms
    batch latency p50 = 47 ms
    batch latency p60 = 47 ms
    batch latency p70 = 52 ms
    batch latency p80 = 57 ms
    batch latency p90 = 58 ms
    batch latency p100 = 21847 ms
    transfer latency p00 = 5 ms
    transfer latency p10 = 4123 ms
    transfer latency p20 = 18843 ms
    transfer latency p30 = 41475 ms
    transfer latency p40 = 73113 ms
    transfer latency p50 = 118810 ms
    transfer latency p60 = 172811 ms
    transfer latency p70 = 220026 ms
    transfer latency p80 = 272064 ms
    transfer latency p90 = 321065 ms
    transfer latency p100 = 361967 ms
    
    # benchmark results after
    1223 batches in 318.06 s
    load offered = 1000000 tx/s
    load accepted = 31440 tx/s
    batch latency p00 = 5 ms
    batch latency p10 = 37 ms
    batch latency p20 = 42 ms
    batch latency p30 = 42 ms
    batch latency p40 = 47 ms
    batch latency p50 = 47 ms
    batch latency p60 = 47 ms
    batch latency p70 = 48 ms
    batch latency p80 = 52 ms
    batch latency p90 = 57 ms
    batch latency p100 = 18951 ms
    transfer latency p00 = 5 ms
    transfer latency p10 = 4206 ms
    transfer latency p20 = 19490 ms
    transfer latency p30 = 41898 ms
    transfer latency p40 = 72858 ms
    transfer latency p50 = 112216 ms
    transfer latency p60 = 153890 ms
    transfer latency p70 = 195212 ms
    transfer latency p80 = 234889 ms
    transfer latency p90 = 272316 ms
    transfer latency p100 = 308050 ms
    ```

Commit:
lsm: implement move table optimization

When the table can be moved directly between levels
without reading values from the iterators, do so and
skip the rest of the compaction process

jorangreef commented:
Nice work @kprotty.

What is the batch size used per transfers batch here?

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -315,6 +317,25 @@ pub fn CompactionType(
+                compaction.merge_done = true;
+                compaction.status = .done;
+                callback(compaction);
+                return;
```
What else can we possibly assert here? For example, compaction range or table count?

jorangreef replied:
It would also be better if we didn't depend on the optionality of `grid_reservation` to indicate `move_table`, so that we can make the intention more clear, and in the positive.

For example, perhaps the `move_table` boolean could be a function of the compaction state, instead of introducing a local constant as we do above. This would then also allow us to assert when `grid_reservation` does exist that we are not moving the table. i.e. Both positive and negative branches for both cases.

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -315,6 +317,25 @@ pub fn CompactionType(
+                compaction.manifest.update_table(level_a, snapshot_max, &table_a);
+                assert(table_a.snapshot_max == snapshot_max);
+
+                table_a = compaction.level_a_input.?;
```
I suggest using a different variable here so that it is more obvious that the original `table_a` was modified.

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -315,6 +317,25 @@ pub fn CompactionType(
+                const level_b = compaction.level_b;
+                const level_a = level_b - 1;
+
+                var table_a = compaction.level_a_input.?;
```
You can assert that `table_a.snapshot_max >= snapshot_max`.

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -315,6 +317,25 @@ pub fn CompactionType(
+                assert(table_a.snapshot_max == snapshot_max);
+
+                table_a = compaction.level_a_input.?;
+                compaction.manifest.insert_table(level_b, &table_a);
```
How does this prevent `remove_invisible_tables()` from cleaning up the level-A table (and all of its blocks, which are still in use) at the end of the compaction?

kprotty replied:
Good point, it doesn't. Could we use `TableInfo.flags` to denote "moved" instead of "removed" when purging invisible tables? Thinking of implementing move_table entirely in Manifest.

kprotty replied:
That, or remove the table right then and there (inside Manifest) instead of waiting for `remove_invisible_tables` at the end of Compaction.

sentientwaffle replied:
Implementing `move_table` in `Manifest` makes sense to me (in fact I think that's where it originally lived).


> That, or remove the table right then and there (inside Manifest) instead of waiting for remove_invisible_tables at the end of Compaction.

There is a comment in `tree.zig`:

```
            // All compactions have finished for the current half-bar.
            // We couldn't remove the (invisible) input tables until now because prefetch()
            // needs a complete set of tables for lookups to avoid missing data.
```

i.e. if you move the table during beat number 1, we will keep querying the table (in its original location/level) in the next beat.

sentientwaffle replied:
I think a table flag is a decent solution, though it is a little weird since the table flags get persisted, but this particular flag would never be persisted.

kprotty replied:
Instead of flags, Im thinking of using a separate buffer/hashmap of `TableInfo.address`

Commit:
lsm: make move_table decision a function of Compaction

This localizes the asserts and makes code easier to follow
without reading the comments.

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -250,22 +253,33 @@ pub fn CompactionType(
 
-            compaction.iterator_a.start(iterator_a_context, iterator_a_io_callback);
-            compaction.iterator_b.start(iterator_b_context, iterator_b_io_callback);
+        /// Returns true if the compaction must use the move table optimization. 
```
There is another complication to move-table that is not yet implemented.

Consider the case of "append-only" trees (i.e. insert-only trees which are keyed by `timestamp`, such as the `Transfer` object table). If we always append keys, then tables never overlap. And if tables never overlap, then we always use move table.

The problem with that is: those tables may be any size. If they are quite small, then the space amplification is high. There are only a limited number of tables too, because of metadata overhead.

The solution is at the last level to coalesce adjacent (but nonoverlapping) tables (instead of a simple move-table).

sentientwaffle replied:
> The solution is at the last level to coalesce adjacent (but nonoverlapping) tables (instead of a simple move-table).

Actually we probably want to be more aggressive than just doing this for the last level. Maybe based on how full the tables are instead.

kprotty replied:
Would the coalescing decision be done in `Manifest.compaction_table`? I assume it would return a range of multiple tables (`range.table_count > 1`) by defaulting not hitting the move-table Compaction paths.

sentientwaffle replied:
Yeah, that sounds good.
Btw, we can include "number of values in the table" to the `TableInfo` so that we know when tables can be coalesced.
(Also keep in mind that the table size can differ between levels, see https://github.com/tigerbeetledb/tigerbeetle/issues/486).

jorangreef replied:
> Actually we probably want to be more aggressive than just doing this for the last level. Maybe based on how full the tables are instead.

Let's hold off on coalescing middle levels for now.

For the last level, coalescing converts a table of normal size (according to the tree) to a table of much larger size (8x bigger), but this is fine since the last level doesn't compact into further levels. Making the table bigger like this doesn't increase the number of tables involved in subsequent compactions, since there aren't any compactions from out of the last level. This is also enough to solve our metadata overhead problem, because the last level has ~90% of the data.

However, for interim levels, of course we can't coalesce to make the table bigger than the normal table size, since this would increase the number of overlapping tables involved in subsequent tables, and vary compaction latencies.

Where coalescing would help for interim levels, is where tables are not being saturated as we expect, up to their normal table size. But I think let's first explore this and get a feeling for how many tables in interim levels are grossly underweight. And then see if we can't prevent or improve that in the first place, before we try to fix it up with coalescing.

Coalescing within a level, ideally also requires the presence of an adjacent table, where both tables together are still within the normal table size. It would be good to get a feeling for how many underweight tables we have typically side by side, if this optimization would pay off. Otherwise, if we coalesce an underweight table with a full weight table, do we end up in a better position?

sentientwaffle replied:
> However, for interim levels, of course we can't coalesce to make the table bigger than the normal table size, since this would increase the number of overlapping tables involved in subsequent tables, and vary compaction latencies.

For object trees (which are append-only) and low-cardinality secondary indexes (which have a handful of long runs of strictly increasing keys because of the timestamp suffix), it doesn't increase the number of overlapping tables, because tables never / rarely overlap (respectively).

> Where coalescing would help for interim levels, is where tables are not being saturated as we expect, up to their normal table size.

@jamii measured this yesterday and it impacts almost all tables, which is specifically why I think we need to coalesce before reaching the last level. Otherwise we are paying a massive space amplification cost — even when data (finally) starts reaching the last level, all previous levels will always remain full of mostly-empty tables.

> Otherwise, if we coalesce an underweight table with a full weight table, do we end up in a better position?

This is why I suggest tracking the table's `value_count` in the `TableInfo` — we can tell in advance whether or not coalescing is an improvement, and use that to decide whether to trigger "coalesce-compaction" or "move-table-compaction". e.g. if we have a table that is at 1/8 capacity, and it will have an adjacent (nonoverlapping) table at 1/4 capacity, we are guaranteed that the coalesced table will be at 3/8 capacity.

Commit:
lsm: Implement move_table in Manifest

The previous approach for move_table would call `update_table(level_a)`
then `insert_table(level_b)`. The next call to `remove_invisible_tables()`
would see the invisible snapshot from `update_table(level_a)` and delete
the blocks for the table's address even though they're still referenced
in `level_b`.

This moves the code into Manifest while staying mostly the same.
Now, the updated `level_a` table is tagged as "moved" using its flags.
When `remove_invisible_tables()` is called, it deletes the table from
the manifest_level as usual but doesn't delete the blocks of "moved" tables.

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -250,22 +253,33 @@ pub fn CompactionType(
-                .direction = .ascending,
-                .table_info_callback = iterator_b_table_info_callback,
-            };
+            // Start the iteraters if we're not only moving one table.
```
```suggestion
            // Start the compaction iterators if we can't move the table directly between levels.
```

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -453,6 +486,9 @@ pub fn CompactionType(
             assert(!compaction.merge_done);
+            assert(!compaction.move_table());
+
+            const grid_reservation = compaction.grid_reservation.?;
```
@kprotty What is the reason for caching this variable here? Line length?

Per TigerStyle, it's better to avoid introducing a variable into the scope like this, because then there's more chance that something is misused. We've had bugs like this before.

Instead, shrink the scope, by continuing to use `grid_reservation.?` inline as before.

kprotty replied:
I moved it to be a checked access at the start to show that its active for the entire function and that it's immutable. I'll change it back.

jorangreef commented on src/lsm/manifest.zig:
```diff
@@ -219,14 +229,41 @@ pub fn ManifestType(comptime Table: type, comptime Storage: type) type {
+
+            // Update the table in level a, marking it explicitely as "moved" to prevent its blocks
+            // from being collected as they're now referenced by the inserted table in level b.
+            manifest_level_a.set_snapshot_max(snapshot, .{ .moved = true }, table);
```
I don't think we should be tracking `.moved` permanently as a flag in the Manifest, simply in order to decide whether to collect blocks after a compaction. The former is long-lived, but this decision is short-lived. We also then don't need to introduce the Flags type yet.

kprotty replied:
What is an alternative way to ensure that "moved" tables with updated snapshots don't get marked as removed to the `ManifestLog` during `remove_invisible_tables()`? The intrusive approach seemed simplest, but we could instead keep a separate hashmap/list that's checked in `remove_invisible_tables()`. Issue there is that the capacity would have to be adjusted with any manifest access changes to Compaction or size changes to higher components like Tree and Groove.

kprotty replied:
Ended up implementing it using a separate buffer. It takes advantage of there only being one moved table for each level until a matching call to `remove_invisible_tables()`.

kprotty committed:
lsm: only reset Compactions after finish accessing them

There was a possible footgun here of compaction.reset()
writing over the fields used/passed into `remove_invisible_tables`.

kprotty committed:
lsm: track moved tables with HashMap instead of Flags

Instead of defining TableInfo.flags to implement move_table,
the tables that are moved can be tracked with a separate buffer.
In this case, a set holding the TableInfo's address.

This also fixes the bug of move_table appending the updated table
insert of level A to manifest log. Per the manifest log docs, it should
only append the insert to level B to move (never the update to level A).
When the log is replaced in LIFO order by open(), it will see the table in
level B first and ignore the updated/moved table in level A.

kprotty committed:
lsm: tiger-style updates

Commit:
lsm: track moved tables per level

The previous solution used a hashset that was shared between
all levels in the manifest. This resulted in footguns on assertions
when multiple levels moved tables between each other.

Since only one table is moved per level between each call to
`remove_invisible_tables`, we can just represent that as an array
of table addresses per level for simplicity.

kprotty committed:
lsm: zig fmt

kprotty committed:
lsm: dont assume remove_invisible_tables on Manifest deinit

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -584,7 +619,7 @@ pub fn CompactionType(
             // TODO(Beat Pacing) This should really be where the compaction callback is invoked,
             // but currently that can occur multiple times per beat.
-            compaction.grid.forfeit(compaction.grid_reservation);
+            if (compaction.grid_reservation) |reservation| compaction.grid.forfeit(reservation);
```
`if (!compaction.move_table()) compaction.grid.forfeit(compaction.grid_reservation.?);`

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -223,8 +223,11 @@ pub fn CompactionType(
-                .grid_reservation = grid.reserve(range.table_count * Table.block_count_max).?,
+                .grid_reservation = blk: {
+                    // Don't reserve these when we just move the table to the next level.
+                    if (range.table_count == 1 and table_a != null) break :blk null;
```
It is awkward that this condition is both here and `move_table()`.
What about:
- `const CompactionStrategy = enum { table_move, table_merge }`
- `strategy: CompactionStrategy` field on the `Compaction`
- s/`.move_table()`/`.strategy == .table_move`
- `const strategy = if (range.table_count == 1 and table_a != null) .table_move else .table_merge;` just before the `compaction.* = ...`, and pass it into the new compaction.
- Use the `strategy` variable when computing the `grid_reservation` too.

kprotty replied:
Will do. Ironically, I had something like this prototyped before (`const Strategy = union(enum) { move_table, merge_tables: Grid.Reservation }`) but instead opted for `fn move_table()` to have less code changes.

sentientwaffle commented on src/lsm/manifest.zig:
```diff
@@ -130,6 +130,8 @@ pub fn ManifestType(comptime Table: type, comptime Storage: type) type {
 
         manifest_log: ManifestLog,
 
+        moved_table_addresses: [constants.lsm_levels]?u64,
```
Since like `levels` this is keyed by the level index, can you move this line up (to follow `levels: [...`).

sentientwaffle replied:
In `fn checkpoint()` can you add a `for (moved_table_addresses) |address| assert(address == null);`. (It's not part of this diff so my comment is all the way over here.)

Maybe in `compaction_range` and `compaction_table` too?

sentientwaffle commented on src/lsm/manifest.zig:
```diff
@@ -154,6 +156,7 @@ pub fn ManifestType(comptime Table: type, comptime Storage: type) type {
                 .node_pool = node_pool,
                 .levels = levels,
                 .manifest_log = manifest_log,
+                .moved_table_addresses = [_]?u64{null} ** constants.lsm_levels,
```
I would probably define this along with the field declaration (like we do for the callbacks), but nbd either way.

sentientwaffle commented on src/lsm/tree.zig:
```diff
@@ -877,14 +877,14 @@ pub fn TreeType(comptime TreeTable: type, comptime Storage: type, comptime tree_
                     .idle => assert(tree.table_immutable.free),
                     .processing => unreachable,
                     .done => {
-                        tree.compaction_table_immutable.reset();
-                        tree.table_immutable.clear();
```
Why move these lines?

kprotty replied:
These were moved after `remove_invisible_tables()` as the function used the Compaction fields but it seemed like a footgun where `compaction.reset()` could change/invalidate the fields.

sentientwaffle replied:
Oh, good catch!

sentientwaffle commented on src/lsm/tree.zig:
```diff
@@ -897,7 +897,6 @@ pub fn TreeType(comptime TreeTable: type, comptime Storage: type, comptime tree_
                     .idle => {}, // The compaction wasn't started for this half bar.
                     .processing => unreachable,
                     .done => {
-                        context.compaction.reset();
```
Why move this line?

kprotty committed:
lsm: explicit Compaction.strategy

The usage and condition between using the Grid and
MergeIterator vs using Manifest.move_table becomes
clearer with an explicit compaction strategy.

kprotty committed:
lsm: verify no pending moved tables in Manifest

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -584,7 +622,10 @@ pub fn CompactionType(
+            if (compaction.strategy == .table_merge) {
+                compaction.grid.forfeit(compaction.grid_reservation.?);
+                compaction.grid_reservation = null;
+            }
```
Just to make it extra-clear that there is no reservation to forfeit:

```suggestion
            } else {
                assert(compaction.grid_reservation == null);
            }
```

kprotty committed:
lsm: extra assert on grid_reservation

sentientwaffle commented:
LGTM. @jorangreef did you want to give it another look before merge?

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -82,6 +82,11 @@ pub fn CompactionType(
+        const CompactionStrategy = enum {
+            table_move,
+            table_merge,
+        };
```
This is better.

Nit: Reorder the enums alphabetically, but also in the order of most frequent use. So `merge` before `move`. This is a bit safer if the enum is ever messed up, then we won't move but rather merge.

jorangreef replied:
We could also drop the `table_` prefixes here I think. It will read nicer than further down whenever we assert on `strategy`. Also because when we merge, we merge multiple tables.

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -207,9 +214,13 @@ pub fn CompactionType(
             const drop_tombstones = manifest.compaction_must_drop_tombstones(level_b, range);
             assert(drop_tombstones or level_b < constants.lsm_levels - 1);
 
+            const strategy: CompactionStrategy =
```
Does the type need to be prefixed with `Compaction`?

jorangreef replied:
Seeing as we're already in the context of `CompactionType` and seeing that we also have things like `Status`?

kprotty replied:
Its needed for the type inference on `.table_move` in the `if` expression. Another alternative is fully spelling out `CompactionStrategy.table_move` instead the anonymous enum variant; allows for type inference on the `const strategy =` then.

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -250,22 +264,22 @@ pub fn CompactionType(
-            compaction.iterator_a.start(iterator_a_context, iterator_a_io_callback);
-            compaction.iterator_b.start(iterator_b_context, iterator_b_io_callback);
+            // Start the compaction iterators if we need to merge tables between levels.
+            if (compaction.strategy == .table_merge) {
```
This is potentially dangerous, if another enum is added, since we don't exhaustively switch.

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -250,22 +264,22 @@ pub fn CompactionType(
+            // Start the compaction iterators if we need to merge tables between levels.
+            if (compaction.strategy == .table_merge) {
+                const iterator_b_context = .{
+                    .grid = grid,
```
Looks like we need to run `zig fmt`?

If so, @kprotty can you setup your editor or environment to automate this for you, if you haven't already? Then we can automate this away in future so it doesn't recur.

kprotty replied:
`zig/zig fmt` seems to report that it's ok (maybe `zig fmt` refers to a later version of zig). Configured it in editor either way.

jorangreef replied:
It was ok, but it looked like `zig fmt` hadn't been run, because the whitespace was different. Format would have corrected that I think.

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -287,6 +301,7 @@ pub fn CompactionType(
             assert(compaction.status == .processing);
             assert(compaction.callback != null);
             assert(!compaction.merge_done);
+            assert(compaction.strategy == .table_merge);
```
For example, on line 303 we say only `merge_done` but then on line 304 we say `table_merge`.

Seems like the enum could be just `.merge` and it would reuse the "human grammer", the reader's internal parsing that they just did on line 303.

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -315,6 +330,26 @@ pub fn CompactionType(
             assert(compaction.io_pending == 0);
             assert(!compaction.merge_done);
 
+            // Move table optimization
```
We can drop this comment since it's echoing the if condition mostly. Otherwise, we should provide more detail and especially motivation in the comment.

Also, comments always end with punctuation (`.` or `:`).

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -315,6 +330,26 @@ pub fn CompactionType(
+                const level_a = level_b - 1;
+
+                var table_a = compaction.level_a_input.?;
+                assert(table_a.snapshot_max >= snapshot_max);
```
Is there a potential off-by-one in the `>=` condition here?

jorangreef replied:
CC'ing @sentientwaffle also to double-check this.

sentientwaffle replied:
Do you mean that this assertion should be stricter, or less strict?
I think it is correct as-is (i.e. not too strict), but I'm not sure whether it could be stricter.

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -315,6 +330,26 @@ pub fn CompactionType(
+                compaction.status = .done;
+                callback(compaction);
+                return;
+            }
```
What do we expect in the `else` branch here?

jorangreef replied:
For example, strategy then has to be `merge`.

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -584,7 +622,13 @@ pub fn CompactionType(
+                .table_move => assert(compaction.grid_reservation == null),
+                .table_merge => {
+                    compaction.grid.forfeit(compaction.grid_reservation.?);
+                    compaction.grid_reservation = null;
```
Should we reset to null here? If so, do we need to clean up other state? I'm not sure that we should set to null.

Or otherwise, it's as if we're using the `grid_reservation` also as a way to assert idempotency?

kprotty replied:
Viewed it as a way to assert other functions aren't being called incorrectly (similar to setting `.callback = null`)

sentientwaffle replied:
> Viewed it as a way to assert other functions aren't being called incorrectly (similar to setting .callback = null)

Perfect; I agree that resetting `grid_reservation` immediately after forfeit is a good practice. It guards against accidentally "leaking" a reservation between compactions.

jorangreef replied:
Yes, agreed that it's a guard. I was wondering though whether we should use it as an _implicit_ guard (as we do here), or whether we should have an _explicit_ guard for this purpose?

@sentientwaffle agreed also with your argument around accidentally "leaking" a reservation across compactions.

That settles it. Let's keep using `null`.

jorangreef commented on src/lsm/manifest.zig:
```diff
@@ -222,11 +225,46 @@ pub fn ManifestType(comptime Table: type, comptime Storage: type) type {
+            const manifest_level_a = &manifest.levels[level_a];
+            const manifest_level_b = &manifest.levels[level_b];
+            if (constants.verify) {
+                assert(manifest_level_a.contains(table));
+                assert(!manifest_level_b.contains(table));
```
Nice!

jorangreef commented on src/lsm/manifest.zig:
```diff
@@ -222,11 +225,46 @@ pub fn ManifestType(comptime Table: type, comptime Storage: type) type {
+                assert(!manifest_level_b.contains(table));
+            }
+
+            // Insert the table into level b and append insert changes to the manifest log.
```
```suggestion
            // Insert the table into level A and append insert changes to the manifest log.
```

Same in the comments below, we can capitalize `A` or `B`.

jorangreef commented on src/lsm/manifest.zig:
```diff
@@ -222,11 +225,46 @@ pub fn ManifestType(comptime Table: type, comptime Storage: type) type {
+            // To move a table w.r.t manifest log, the previous level's table should not be removed.
+            // When replaying the log from open(), inserts are processed in LIFO order while
+            // duplicates are ignored. This means the table will only appear in level_b as moved.
+            manifest_level_a.set_snapshot_max(snapshot, table);
```
Is it possible to hoist this above the "insert to level_b"? So we deal with level A and then level B in order?

Is there a bug here where we might have two copies of the table in both levels with the same snapshot? CC'ing @sentientwaffle

sentientwaffle replied:
Currently the order is important because we need to ensure that `manifest_level_b.insert_table` happens before `manifest_level_a.set_snapshot_max` modifies the table's snapshot.

But that is subtle — maybe we should instead copy `table` to the stack (`s/table/table_a` + `const table_b = table_a.*`), then handle them in order (`manifest_level_a.set_snapshot_max(snapshot, table_a);` then `manifest_level_b.insert_table(manifest.node_pool, table_b);`.

jorangreef replied:
Thanks, that would be good—or we could add a comment to explain why the order is important?

Also, is it possible that the two tables share the same snapshot value, to be seen as duplicates (if level b was also probed in the read path by a subsequent snapshot)?

sentientwaffle replied:
> Also, is it possible that the two tables share the same snapshot value, to be seen as duplicates (if level b was also probed in the read path by a subsequent snapshot)?

Just to double-check, you mean that a subsequent query overlaps with the table in both level A and B?
Maybe we should remove the table from `manifest_level_a` instead of just setting its snapshot. I think that would also negate the need for `moved_table_addresses`.

kprotty replied:
Would removing from `manifest_level_a` during `move_table` still invalidate a subsequent `StateMachine.prefetch` in the middle of a half-measure? https://github.com/tigerbeetledb/tigerbeetle/pull/463#discussion_r1113459688

sentientwaffle replied:
By avoiding the snapshot update I think we would avoid triggering `remove_invisible_tables`.

kprotty replied:
It would be called on level A at the end (since the Compaction was started), but it wont have anything to remove. IIUC, the primary difference would be that, after a tree.compact() in the middle of a half-measure, prefetch() can observe that the table either still exists in level A with the updated snapshot or doesn't exist in level A at all but atomically now exists in level B.

sentientwaffle replied:
> prefetch() can observe that the table either still exists in level A with the updated snapshot or doesn't exist in level A at all but atomically now exists in level B.

Your wording here confuses me. My thinking is:

- When we do a move-table, we don't update the snapshot at all.
- We atomically move the table (in memory, i.e. `ManifestLevel`s) from level A to level B.

After the atomic move, the table is still available, just from a different level.

jorangreef replied:
> After the atomic move, the table is still available, just from a different level.

Yes, this feels like an easier mental model to me. I can't think of any counter examples where it could go wrong.

jorangreef commented on src/lsm/manifest.zig:
```diff
@@ -222,11 +225,46 @@ pub fn ManifestType(comptime Table: type, comptime Storage: type) type {
+            // Finally, the table address is recorded so that it will not be removed from the
+            // manifest log during `remove_invisible_tables`.
+            assert(manifest.moved_table_addresses[level_a] == null);
+            manifest.moved_table_addresses[level_a] = table.address;
```
If I understand correctly, I think it would be better to use a different optional struct here, containing `level` and `address`. This ensures that only one move table can exist until `remove_invisible_tables()` is called.

kprotty replied:
It needs multiple entries as we use the same `tree.manifest` to compact multiple `[lsm_levels/2]CompactionTable`s (+ `CompactionTableImmutable`) where each of them could do a `tree.maniefst.move_table()` under the same half-beat (`remove_invisible_tables()` is called when each half-measure ends).

jorangreef replied:
Thanks for the explanation @kprotty !

Would be fantastic to be able to remove this structure completely as @sentientwaffle suggested.

jorangreef commented on src/lsm/manifest_level.zig:
```diff
@@ -86,7 +86,11 @@ pub fn ManifestLevelType(
+            level: *Self,
+            snapshot: u64,
+            table: *TableInfo,
+        ) void {
```
This is still under 100 columns I think?

Why do we add the trailing comma after the `*TableInfo` arg, since there isn't a need to reflow right?

kprotty replied:
whoops, artifact from old `TableInfo.flags` implementation

jorangreef reviewed (changes requested):
Nice work @kprotty.

Commit:
lsm: address some review changes & code cleanup

Commit:
lsm: remove moved-table immediately

To move a table, its snapshot is updated and its address is recorded
to avoid appending a "remove" entry to the ManifestLog when the
table is removed from ManifestLevel during `remove_invisible_tables`.

This is simplified to instead remove the table from the ManifestLevel
directly inside `move_table` without updating its snapshot and waiting
for `remove_invisible_tables` to collect it. This avoids having to track
a separate array of TableInfo addresses to check against later.

Commit:
lsm: update table snapshot before removal in move-table

`ManifestLevel.remove_table` relies on the table being invisible to
the given snapshot before it can remove it. This means we need to
follow the normal flow of updating its `snapshot_max` first before removal.

Commit:
lsm: simplify manifest.move_table()

The table's snapshot doesn't need to be updated before
removing it from the level. Instead, the level implements
a separate remove() function that doesn't check for snapshot
invisibility.

Also, remove from level A before inserting to level B to further
emphasize its subtle ordering.

kprotty committed:
lsm: update old uses of ManifestLevel.remove_table in test

sentientwaffle commented on src/lsm/manifest_level.zig:
```diff
@@ -117,9 +117,41 @@ pub fn ManifestLevelType(
         /// Remove the given table from the ManifestLevel, asserting that it is not visible
         /// by any snapshot in `snapshots` or by `lsm.snapshot_latest`.
-        pub fn remove_table(
+        pub fn remove_table_invisible(
```
Can just call `remove_table` and assert that the table in question is invisible?

sentientwaffle replied:
May need to remove `assert(table.visible(lsm.snapshot_latest));` from `remove_table` for it to work.

kprotty committed:
lsm: share code between ManifestLevel.remove_table* functions

snapshot handling is done external to SetAssociativeCache removal

sentientwaffle commented on src/lsm/manifest_level.zig:
```diff
@@ -117,16 +117,34 @@ pub fn ManifestLevelType(
+        /// Remove the given table from the level assuming it's visible to `lsm.snapshot_latest`.
+        /// Returns the same, unmodified table passed in to differentiate itself from 
+        /// remove_table_invisible and guard against using the wrong function.
+        pub fn remove_table(
```
Maybe call this `remove_table_visible`, and then `s/remove/remove_table`?

kprotty committed:
lsm: ManifestLevelt remove_table_visible

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -223,8 +234,11 @@ pub fn CompactionType(
+                .grid_reservation = switch (strategy) {
+                    // Don't reserve these when we just move the table to the next level.
+                    .move => null,
+                    .merge => grid.reserve(range.table_count * Table.block_count_max).?,
```
Let's use consistent order everywhere, also to follow the typical logical order of use (and the reason we re-ordered the enum): `merge` always before `move`.

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -315,30 +334,49 @@ pub fn CompactionType(
+                    const level_b = compaction.level_b;
+                    const level_a = level_b - 1;
+
+                    const table_a = compaction.level_a_input.?;
```
Looks like this is copying the `TableInfo` to the stack? Can't we take a constant pointer here instead?

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -315,30 +334,49 @@ pub fn CompactionType(
+
+                    compaction.merge_done = true;
+                    compaction.status = .done;
+                    callback(compaction);
```
What else do we need to do in terms of clean up here before calling the callback?

kprotty replied:
I believe those two variables are the primary updates needed before completing. It's mimicked in `cpu_merge_finish` and `reset()` checks for their clean up.

jorangreef commented on src/lsm/manifest.zig:
```diff
@@ -222,11 +222,40 @@ pub fn ManifestType(comptime Table: type, comptime Storage: type) type {
+            const removed = manifest_level_a.remove_table_visible(manifest.node_pool, table);
+            assert(table.equal(removed));
+
+            // Then, insert the table into the level B and append these changes to the manifest log.
```
```suggestion
            // Then, insert the table into level B and append these changes to the manifest log.
```

jorangreef commented on src/lsm/manifest.zig:
```diff
@@ -222,11 +222,40 @@ pub fn ManifestType(comptime Table: type, comptime Storage: type) type {
+            // LIFO order and duplicates are ignored. This means the table will only be replayed in
+            // level B instead of the old one in level A.
+            manifest_level_b.insert_table(manifest.node_pool, table);
+            manifest.manifest_log.insert(@intCast(u7, level_b), table);
```
What do we want to verify at the end of this function?

jorangreef replied:
Where I'm going with this is that we should verify at the end, the inverse of what we also verify at the beginning.

jorangreef commented on src/lsm/manifest_level.zig:
```diff
@@ -117,16 +117,34 @@ pub fn ManifestLevelType(
             level.table_count_visible -= 1;
         }
 
+        /// Remove the given table from the level assuming it's visible to `lsm.snapshot_latest`.
+        /// Returns the same, unmodified table passed in to differentiate itself from 
+        /// remove_table_invisible and guard against using the wrong function.
```
Great comment here, and good idea to differentiate function signatures for safety.

jorangreef commented on src/lsm/manifest_level.zig:
```diff
@@ -701,7 +716,7 @@ pub fn TestContext(
 
                 if (to_remove.items.len > 0) {
                     for (to_remove.items) |*table| {
-                        context.level.remove_table(
+                        context.level.remove_table_invisible(
```
Since the function signature is the same, have you double-checked that we've updated all necessary call sites in the code base to use `remove_table_invisible` instead?

kprotty replied:
I believe so. CI caught some here in the tests previously. It helps that one version accepts snapshots while another doesn't.

Commit:
lsm: compaction/manifest cleanups

Commit:
lsm: fix table address typo

kprotty committed:
lsm: zig fmt

kprotty committed:
lsm: assert table location after move_table()

kprotty commented:
This should be good to go now. Could you give it a final look over again please? @sentientwaffle
